README
Padre-Plugin-Debugger
=====================

This is a very early pre-release of a perl debugger for padre.


INSTALLATION
============

Padre has to be installed before installing this module.
To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

AUTHOR
======

Peter Makholm, <peter@makholm.net>


COPYRIGHT & LICENSE
===================

Copyright (c) 2009 Peter Makholm, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for Padre-Plugin-Debugger

0.3     2009-06-12
       Call the interpreter according to the documents 'Run Parameters' from
       Edit -> Preferences -> Run Parameters.
       Reordered the menu.
       A lot of other minor details.

0.2     2009-05-12
    Forgot to write a changelog,

0.1     2009-05-06
    Working debugger for single file debugging.

0.0_1   2009-05-05
    Initial very rough version


Build.PL
#
# This file is part of Padre::Plugin::Debug.
# Copyright (c) 2009 Peter Makholm, all rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
#

use strict;
use warnings;

use Module::Build;
my $BuildClass = "Module::Build";

eval {
    require Padre::PluginBuilder;
    $BuildClass = "Padre::PluginBuilder";
};

my $builder = $BuildClass->new(
    module_name         => 'Padre::Plugin::Debugger',
    license             => 'perl',
    dist_author         => 'Peter Makholm <peter@makholm.net>',
    dist_version_from   => 'lib/Padre/Plugin/Debugger.pm',
    build_requires => {
        'Test::More' => 0,
    },
    requires => {
        'Devel::ebug'           => '0.49',
        'File::Basename'        => 0,
        'File::Spec::Functions' => 0,
        'Padre'                 => '0.28',
        'Padre::Plugin'         => '0.28',
        'parent'                => 0,
    },
    meta_merge => {
        resources => {
            repository => 'http://github.com/pmakholm/padre-plugin-debugger/tree/master',
        }
    }

);

$builder->create_build_script();


lib/Devel/ebug/Padre.pm
package Devel::ebug::Padre;

use String::Koremutake;
use IO::Socket::INET;
use Class::Accessor::Chained::Fast;

use base (Devel::ebug);

__PACKAGE__->mk_accessors(qw(
    interpreter interpreter_args program_args));

our $VERSION = "0.49";

# This is a copy paste from Devel::ebug, except that the $command variable is
# more configurable

sub load {
  my $self = shift;
  my $program = $self->program;
  my $interpreter = $self->interpreter || $^X;
  my $interpreter_args = defined( $self->interpreter_args ) ?  $self->interpreter_args : "-Ilib";
  my $program_args = defined( $self->program_args ) ? $self->program_args : "";

  # import all the plugins into our namespace
  do { eval "use $_ " } for $self->plugins;

  my $k = String::Koremutake->new;
  my $rand = int(rand(100_000));
  my $secret = $k->integer_to_koremutake($rand);
  my $port   = 3141 + ($rand % 1024);

  $ENV{SECRET} = $secret;
  my $command = join " ", $interpreter, $interpreter_args, "-d:ebug::Backend", $program, $program_args;
#  warn "Running: $command\n";
  my $proc = Proc::Background->new(
    {'die_upon_destroy' => 1},
    $command
  );
  croak(qq{Devel::ebug: Failed to start up "$program" in load()}) unless $proc->alive;
  $self->proc($proc);
  $ENV{SECRET} = "";

  # try and connect to the server
  my $socket;
  foreach (1..10) {
    $socket = IO::Socket::INET->new(
      PeerAddr => "localhost",
      PeerPort => $port,
      Proto    => 'tcp',
      Reuse      => 1,
      ReuserAddr => 1,
    );
    last if $socket;
    sleep 1;
  }
  die "Could not connect: $!" unless $socket;
  $self->socket($socket);
 
  my $response = $self->talk({
    command => "ping",
    version => $VERSION,
    secret  => $secret,
  });
  my $version = $response->{version};
  die "Client version $version != our version $VERSION" unless $version eq $VERSION;
  $self->basic; # get basic information for the first line
}

1;


lib/Padre/Plugin/Debugger/Wx/Menu.pm
package Padre::Plugin::Debugger::Wx::Menu;

use Padre::Wx::Menu ();
use base 'Padre::Wx::Menu';

sub new {
    my $class  = shift;
    my $main   = shift;
    my $plugin = shift;

    # Create the empty menu as normal
    my $self   = $class->SUPER::new(@_);

    # Add additional properties
    $self->{main}   = $main;
    $self->{plugin} = $plugin;

    $self->{start} = $self->Append(
        -1,
        "Start debugger\tShift+Alt+F5",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{start},
        sub { 
            $plugin->start_debugger;
        },
    );

    $self->{stop} = $self->Append(
        -1,
        "Stop debugger",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{stop},
        sub { 
            $plugin->stop_debugger;
        },
    );

    # View sub menu
    my $view_menu = Wx::Menu->new();
    $self->Append(
        -1,
        "View...",
        $view_menu
    );

    $self->{view_stacktrace} = $view_menu->AppendCheckItem(
        -1,
        "Show stacktrace",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{view_stacktrace},
        sub { 
            $plugin->show_stacktrace( $_[1]->IsChecked );
        },
    );

    $self->{view_watches} = $view_menu->AppendCheckItem(
        -1,
        "Show watches",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{view_watches},
        sub { 
            $plugin->show_watches( $_[1]->IsChecked );
        },
    );

    # Run sub menu
    my $run_menu = Wx::Menu->new();
    $self->Append(
        -1,
        "Run...",
        $run_menu
    );

    $self->{continue} = $run_menu->Append(
        -1,
        "Continue\tShift+Alt+C",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{continue},
        sub { 
            $plugin->debug_continue;
        },
    );

    $self->{step} = $run_menu->Append(
        -1,
        "Step\tShift+Alt+S",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{step},
        sub { 
            $plugin->debug_step;
        },
    );

    $self->{next} = $run_menu->Append(
        -1,
        "Next\tShift+Alt+N",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{stop},
        sub { 
            $plugin->debug_next;
        },
    );

    $self->{return} = $run_menu->Append(
        -1,
        "Return\tShift+Alt+R",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{return},
        sub { 
            $plugin->debug_return;
        },
    );


    # Breakpoints and watches
    my $break_menu = Wx::Menu->new();
    $self->Append(
        -1,
        "Breakpoints and watches...",
        $break_menu,
    );

    $self->{breakpoint} = $break_menu->Append(
        -1,
        "Add breakpoint\tShift+Alt+B",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{breakpoint},
        sub { 
            $plugin->debug_breakpoint;
        },
    );

    $self->{breakpoint_cond} = $break_menu->Append(
        -1,
        "Add breakpoint (conditional)\tCtrl+Shift+Alt+B",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{breakpoint_cond},
        sub { 
            $plugin->debug_breakpoint_cond;
        },
    );

    $self->{watch} = $break_menu->Append(
        -1,
        "Add watch\tShift+Alt+W",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{watch},
        sub { 
            $plugin->debug_watch;
        },
    );

    # Main menu
    $self->{eval} = $self->Append(
        -1,
        "Eval expression\tShift+Alt+E",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{eval},
        sub { 
            $plugin->debug_eval;
        },
    );

    $self->AppendSeparator;
    $self->{about} = $self->Append(
         Wx::wxID_ABOUT,
        "About",
    );
    Wx::Event::EVT_MENU(
        $main,
        $self->{about},
        sub { 
            $plugin->show_about;
        },
    );

    return $self;
}

sub refresh {
    my $self       = shift;
    my $plugin     = $self->{plugin};
    my $main       = $self->{main};
    my $document   = $main->current->document;
    my $is_running = $plugin->is_running;

    $self->{start}->Enable( !$is_running );
    $self->{stop}->Enable(  $is_running );

    $self->{$_}->Enable( $is_running )
        for qw( 
            continue step next return
            breakpoint breakpoint_cond watch
            eval
        );

    return 1;
}

1;


lib/Padre/Plugin/Debugger.pm
#
# This file is part of Padre::Plugin::Debugger.
# Copyright (c) 2009 Peter Makholm, all rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
#

package Padre::Plugin::Debugger;

use strict;
use warnings;

use File::Basename qw(fileparse);
use File::Spec::Functions qw(catfile abs2rel rel2abs);
use YAML;

use Padre::Wx;
use Padre::Plugin;

use Padre::Plugin::Debugger::Wx::Menu;

use parent qw(Padre::Plugin);

our $VERSION = "0.3";

# -- Padre API, see Padre::Plugin

sub plugin_name { "Debugger" }

sub padre_interfaces {
    "Padre::Plugin" => 0.28,
}

sub menu_plugins {
    my $self = shift;
    my $main = shift;

    $self->{menu} ||= Padre::Plugin::Debugger::Wx::Menu->new($main,$self);
    $self->{menu}->refresh();

    return ( $self->plugin_name => $self->{menu}->wx );
}

sub plugin_enable {
    my $self = shift;
    my $main = Padre->ide->wx->main;

    $self->{config} = $self->config_read() || {};

    $self->{menu} ||= Padre::Plugin::Debugger::Wx::Menu->new($main,$self);
    $self->{menu}->refresh();

    $self->show_stacktrace( $self->{config}->{stacktrace} );
    $self->show_watches( $self->{config}->{watchbox} );

    return 1;
}

# Public functions

sub is_running {
    my $self = shift;

    return exists $self->{debugger};
}

sub show_about {
    my $self = shift;

    # Generate the About dialog
    my $about = Wx::AboutDialogInfo->new;
    $about->SetName("Padre::Plugin::Debugger");
    $about->SetDescription( <<"END_MESSAGE" );
Padre Perl5 Debugger
END_MESSAGE
    $about->SetVersion( $VERSION );

    # Show the About dialog
    Wx::AboutBox( $about );

    return;
}

sub start_debugger {
    my $self = shift;
    my $main = Padre->ide->wx->main;
    my $doc  = $main->current->document;

    if (exists $self->{debugger}) {
        $main->error("Debugger is allready running");
        return;
    }

    # Copied from Document/Perl.pm
    my $config = Padre->ide->config;

    # Run with the same Perl that launched Padre
    # TODO: get preferred Perl from configuration
    my $file = $doc->filename;
    my $perl = Padre->perl_interpreter;

    # Set default arguments
    my %run_args = (
        interpreter => $config->run_interpreter_args_default,
        script      => $config->run_script_args_default,
    );

    # Overwrite default arguments with the ones preferred for given document
    foreach my $arg ( keys %run_args ) {
        my $type = "run_${arg}_args_" . File::Basename::fileparse($file);
        $run_args{$arg} = Padre::DB::History->previous($type) if Padre::DB::History->previous($type);
    }

    my $dir = File::Basename::dirname($file);
    chdir $dir;
    
    require Devel::ebug::Padre;
    my $ebug = Devel::ebug::Padre->new();

    $ebug->interpreter($perl);
    $ebug->interpreter_args($run_args{interpreter});
    $ebug->program($file);
    $ebug->program_args($run_args{script});

    $ebug->load;

    $self->{debugger} = $ebug;
    $self->update_view;
}

sub stop_debugger {
    my $self = shift;
    my $main = Padre->ide->wx->main;
    my $doc  = $main->current->document;

    unless (exists $self->{debugger}) {
        $main->error("Debugger isn't running");
        return;
    }

    $main->message("Debugger stopped");

    delete $self->{debugger};
}

sub debug_step {
    my $self = shift;
    my $ebug = $self->{debugger};
    my $main = Padre->ide->wx->main;
    my $file = $main->current->document->filename;

    unless (defined $ebug) {
        $main->error("Debugger isn't running");
        return;
    }

    do { $ebug->step } until $self->update_view;
}

sub debug_continue {
    my $self = shift;
    my $ebug = $self->{debugger};
    my $main = Padre->ide->wx->main;

    unless (defined $ebug) {
        $main->error("Debugger isn't running");
        return;
    }

    $ebug->run;
    $self->update_view;
}

sub debug_next {
    my $self = shift;
    my $ebug = $self->{debugger};
    my $main = Padre->ide->wx->main;

    unless (defined $ebug) {
        $main->error("Debugger isn't running");
        return;
    }

    $ebug->next;
    $self->update_view;
}

sub debug_return {
    my $self = shift;
    my $ebug = $self->{debugger};
    my $main = Padre->ide->wx->main;

    unless (defined $ebug) {
        $main->error("Debugger isn't running");
        return;
    }

    $ebug->return;
    $self->update_view;
}

sub debug_eval {
    my $self = shift;
    my $ebug = $self->{debugger};
    my $main = Padre->ide->wx->main;

    unless (defined $ebug) {
        $main->error("Debugger isn't running");
        return;
    }

    my $eval = $main->prompt("Evaluate in debugger", "Please type expression to evaluate", "MY_DEBUGGER_EVAL");

    if ($eval) {
        my $yaml = $ebug->yaml($eval);
        $main->message($yaml, "Result");
    }
    return 1;
}

sub debug_breakpoint {
    my $self = shift;
    my $cond = shift;
    my $ebug = $self->{debugger};
    my $main = Padre->ide->wx->main;
    my $file = $main->current->document->filename;

    $file = $self->ebug_filename( $file );

    unless (defined $ebug) {
        $main->error("Debugger isn't running");
        return;
    }

    my $editor = Padre::Current->editor;
    my $line   = $editor->LineFromPosition($editor->GetCurrentPos);
    my $break  = $ebug->break_point($file, $line + 1, $cond) - 1;

    # Make marker:
    my $red    = Wx::Colour->new("red");

    $editor->MarkerDefine( MarkBreakPoint(), Wx::wxSTC_MARK_ARROW, $red, $red );
    $editor->MarkerAdd( $break, MarkBreakPoint() );

    return 1;
}
    
sub debug_breakpoint_cond {
    my $self = shift;
    my $ebug = $self->{debugger};
    my $main = Padre->ide->wx->main;

    unless (defined $ebug) {
        $main->error("Debugger isn't running");
        return;
    }

    my $cond = $main->prompt("Conditional breakpoint", "Please type condition to break on", "MY_DEBUGGER_BREAK");

    $self->debug_breakpoint($cond);
}

sub debug_watch {
    my $self = shift;
    my $ebug = $self->{debugger};

    my $main = Padre->ide->wx->main;

    unless (defined $ebug) {
        $main->error("Debugger isn't running");
        return;
    }

    my $watch = $main->prompt("Watch expression", "Please type expression to watch", "MY_DEBUGGER_WATCH");

    $self->{watches}->{$watch} = $ebug->eval($watch);
    $ebug->watch_point($watch);

    $self->update_view();
}

sub show_watches {
    my $self = shift;
    my $on   = @_ ? $_[0] ? 1 : 0 : 1;
    my $main = Padre->ide->wx->main;

    # Create pane
    require Padre::Plugin::Debugger::Wx::Watches;
    $self->{watchbox} ||=
        Padre::Plugin::Debugger::Wx::Watches->new($main);
    $self->{watchbox}->set_debugger($self);

    # Update view
    unless ( $on == $self->{menu}->{view_watches}->IsChecked ) {
        $self->{menu}->{view_watches}->Check($on);
    }
    $self->{config}->{watchbox} = $on;

    if ($on) {
        $main->bottom->show( $self->{watchbox} );
    } else {
        $main->bottom->hide( $self->{watchbox} );
    }

    $main->aui->Update;

    $self->config_write( $self->{config} );
    $self->update_view() if $self->is_running;

    return 1;
}
sub show_stacktrace {
    my $self = shift;
    my $on   = @_ ? $_[0] ? 1 : 0 : 1;
    my $main = Padre->ide->wx->main;

    # Create pane
    require Padre::Plugin::Debugger::Wx::StackTrace;
    $self->{stacktrace} ||=
        Padre::Plugin::Debugger::Wx::StackTrace->new($main);
    $self->{stacktrace}->set_debugger($self);

    # Update view
    unless ( $on == $self->{menu}->{view_stacktrace}->IsChecked ) {
        $self->{menu}->{view_stacktrace}->Check($on);
    }
    $self->{config}->{stacktrace} = $on;

    if ($on) {
        $main->right->show( $self->{stacktrace} );
    } else {
        $main->right->hide( $self->{stacktrace} );
    }

    $main->aui->Update;

    $self->config_write( $self->{config} );
    $self->update_view() if $self->is_running;

    return 1;
}

sub goto_frame {
    my $self = shift;
    my $line = shift;

    my @stack = $self->{debugger}->stack_trace;
    my $frame = $stack[$line];
    my $file  = $self->padre_filename( $frame->filename );

    my $main = Padre->ide->wx->main;
    my $id   = $main->find_editor_of_file( $file );
    unless ( defined $id ) {
        my $load = Wx::MessageBox(
	    "Unknown file, Should I load it?",
            "Padre", 
            Wx::wxYES_NO | Wx::wxCENTRE, 
            $main
        );
        return if $load == Wx::wxNO;

        $id = $main->setup_editor( $file );
    }

    $main->on_nth_pane($id);
    Padre::Current->editor->goto_line_centerize($frame->line - 1);

    return 1;
}

# Internal functions

sub MarkBreakPoint { 17 }

sub ebug_filename {
    my $self = shift;
    my $file = shift;
    my $ebug = $self->{debugger};

    my $base  = $ebug->eval("require Cwd; Cwd::cwd;");
    my %known = map { rel2abs($_, $base) => $_ } $ebug->filenames;

    return $known{$file};
}

sub padre_filename {
    my $self = shift;
    my $file = shift;
    my $ebug = $self->{debugger};

    my $base  = $ebug->eval("require Cwd; Cwd::cwd;");
    my %known = map { $_ => rel2abs($_, $base) } $ebug->filenames;

    return $known{$file};
}

sub update_view {
    my $self   = shift;
    my $ebug   = $self->{debugger};
    my $editor = Padre::Current->editor;
    my $main = Padre->ide->wx->main;
    return unless $ebug;
    
    if ( $ebug->finished ) {
        $self->stop_debugger;
        $self->{menu}->refresh();
        return 1;
    }


    # Try to change to right file
    my $file = $self->padre_filename( $ebug->filename );
    if ( $main->current->document->filename ne $file ) {
        my $id = $main->find_editor_of_file( $file );
	return unless defined $id; # Autoload files?

	$main->on_nth_pane($id);
        $editor = Padre::Current->editor;
    }

    # Move to current line
    my $line   = $ebug->line() - 1;
    $editor->goto_line_centerize($line);

    # Update watches
    if ($self->{watchbox}) {
        $self->{watches}->{$_} = $ebug->eval($_) for keys %{ $self->{watches} };

        $self->{watchbox}->clear;
        $self->{watchbox}->AppendText( YAML::Dump( $self->{watches} ) );
    }

    # Update stack trace
    if ($self->{stacktrace}) {
        $self->{stacktrace}->DeleteAllItems;

        my @stack = $ebug->stack_trace;

        $self->{stacktrace}->InsertStringItem( 0, $_->filename. "::" .  $_->line)
            for reverse @stack;
        
        $self->{stacktrace}->SetColumnWidth( 0, Wx::wxLIST_AUTOSIZE );
    }

    # Update output
    my($stdout, $stderr) = $ebug->output;
    $main->output->clear;               # We get the full output each time...
    $main->output->AppendText($stdout);
    $main->output->AppendText($stderr);

    # Refresh menu
    $self->{menu}->refresh();

    return 1;
}

1;

__END__

=head1 NAME

Padre::Plugin::Debugger - Debug Perl code from Padre editor



=head1 SYNOPSIS

    $ padre



=head1 DESCRIPTION

This plugin allows one to debug perl code from within Padre. 

=head1 BUGS

Many! This is an very early alpha release. You did notice the version number,
right?

Please report any bugs or feature requests to C<padre-plugin-debugger at
rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Padre-Plugin-Debugger>. I will
be notified, and then you'll automatically be notified of progress on
your bug as I make changes.



=head1 SEE ALSO

L<Devel::ebug> - The backend debugger.

L<http://github.com/pmakholm/padre-plugin-debugger/tree/master> - Git repository

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Padre-Plugin-Debugger> - Bug tracking

=head1 AUTHOR

Peter Makholm, C<< <peter@makholm.net> >>



=head1 COPYRIGHT & LICENSE

Copyright (c) 2009 Peter Makholm, all rights reserved.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
